### PR TITLE
Masonry block use inner core/image blocks

### PR DESCRIPTION
**Pull request by AnthonyLedesma**
### Description

Masonry Block Refactors.

This is moving the Masonry Gallery away from desandro masonry into a pure CSS solution for layout. 

New block architecture reflects that of the Core Gallery block in that this implementation uses Core Image inner blocks. 

Backward compatible to WP 5.5 by keeping the V1 block around. We conditionally load the V1 block to keep functionality as well as preserve masonry styles in the case of pre 5.6 WordPress.

**AC**
This criteria has been modified - Notes below.
- [x] <del>[ISBAT sort images in Masonry Gallery - Something like ascending, descending by upload date, or random, or append to top or bottom.](https://github.com/godaddy-wordpress/coblocks/issues/1890)</del>
  - This is a complicated feature and should be handled in a future iteration.
- [x] [ISBAT have a variety of image size display options available to users](https://github.com/godaddy-wordpress/coblocks/issues/1756) - (400px, 1200px etc)
- [x] <del>[ISBAT access image replace controls both on image and in block toolbar](https://github.com/godaddy-wordpress/coblocks/issues/1749)</del>
  - Image replace and move controls will not exist presently on the Masonry block. This feature is not mirrored in the new Gallery architecture but may come from the core as the v1 Gallery did.
- [x] [ISBAT transform into other blocks that have gutter controls](https://github.com/godaddy-wordpress/coblocks/issues/1719)
- [x] [ISBAT update image captions from within the media library as well as block editor](https://github.com/godaddy-wordpress/coblocks/issues/1620)
- [x] [ISBAT has a mobile view that displays images in a single column](https://github.com/godaddy-wordpress/coblocks/issues/1538)
- [x] [ISN use jQuery in refactored Masonry Gallery](https://github.com/godaddy-wordpress/coblocks/issues/1365)
- [x] <del>[ISB using the newer size controls and deprecated the old controls](https://github.com/godaddy-wordpress/coblocks/issues/1041)</del>
  - Size controls have been deprecated and removed. I think we should consider the fourth masonry column when the block is set to full width. 

### Screenshots

MasonryCSS

### Types of changes

Major changes discarding all of our old CSS and JS in favor of a mirrored core implemntation using hooks. This new method matches the new paradigm and allows us to refactor other galleries to this paradigm more easily.

### How has this been tested?

WIP

### Checklist:
- [ ] My code is tested
- [ ] My code follows accessibility standards 
- [ ] My code has proper inline documentation 
- [ ] I've included any necessary tests 
- [ ] I've included developer documentation 
- [ ] I've added proper labels to this pull request

### CODE REVIEW

**Comment by jrtashjian** [comment]
Can we change the arrows to point up and down instead of left and right? I think it's a prop we can set.

**Comment by olafleur-godaddy** [comment]
When uploading a PNG image, I have this error

Screen Shot 2021-11-16 at 15 19 42

**Comment by olafleur-godaddy** [comment]
When creating a Masonry block in master, it deprecates without error in your branch, but I don't see my images in the Gallery anymore. Is it normal?

File: src/blocks/gallery-masonry/controls.js [hunk 1]
```diff
@@ -1,28 +0,0 @@
-/**
```
Comment (olafleur-godaddy) [inline src/blocks/gallery-masonry/controls.js:1]: Why is this removed?

File: src/blocks/gallery-masonry/inspector.js [hunk 2]
```diff
@@ -1,124 +0,0 @@
-/**
```
Comment (olafleur-godaddy) [inline src/blocks/gallery-masonry/inspector.js:1]: Same question here. It's not needed anymore?

**Comment by AnthonyLedesma** [comment]
> When creating a Masonry block in master, it deprecates without error in your branch, but I don't see my images in the Gallery anymore. Is it normal?

No, the block is not deprecating as expected.

File: src/blocks/gallery-masonry/controls.js (hunk 1 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/controls.js:1]: Media Filters were applied ad-hoc to the custom inner images that we used in the galleries. We apply the Media Filters by Extension to the core/image block so we already have those controls on the images thus we do not need them here.

Comment (olafleur-godaddy) [reply src/blocks/gallery-masonry/controls.js:1]: Ok, thanks for the explanation!

File: src/blocks/gallery-masonry/inspector.js (hunk 2 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/inspector.js:1]: These controls have been moved to the edit.js file now. I wonder which might be improved readability either to split the code or to have all of the controls within the edit function.

**Comment by AnthonyLedesma** [comment]
> >    When creating a Masonry block in master, it deprecates without error in your branch, but I don't see my images in the Gallery anymore. Is it normal?

> No, the block is not deprecating as expected.

This should be resolved now with a successful deprecation and conversion to innerblocks.

**Comment by olafleur-godaddy** [comment]
You can probably change the minimal version of WP from 5.5 to 5.6 in the readme, as the v2 block API was introduced in WP 5.6

**Comment by olafleur-godaddy** [comment]
When creating a block in master, saving it, loading it in your branch it works great, but next time I do a save, it breaks the block.

**Comment by AnthonyLedesma** [comment]
> When creating a block in master, saving it, loading it in your branch it works great, but next time I do a save and refresh, it breaks the block.

I think I got it this time. Now it is working with markup generated from master and is persisting past save.

**Comment by olafleur-godaddy** [comment]
Nice! Yes, it's deprecating properly now :)

**Comment by olafleur-godaddy** [comment]
When uploading one or multiple images in the Gallery, I have this message show in the console : 
Screen Shot 2021-11-19 at 11 44 50

Do you also have that?

**Comment by AnthonyLedesma** [comment]
The Jest tests for Masonry deprecations only work properly when using the updated block-editor package.

**Comment by AnthonyLedesma** [comment]
Now that I look at these jest errors it seems like the snapshots just need updating. The markup in the snapshots seems to be produced by upstream dependency updates.

**Comment by olafleur-godaddy** [comment]
Not sure if it's expected behaviour, but when adding multiple images at the same time, the gallery reloads a few times (I think the number of images added) before stabilizing.

**Review by olafleur-godaddy — APPROVED** [review]
This is monumental work, good job! :)

**Comment by olafleur-godaddy** [comment]
Whenever you feel ready, you can merge, as the failing tests for OpenTable are not related to your PR :)

**Review by jrtashjian — APPROVED** [review]

**Review by jrtashjian — APPROVED** [review]
Impressive work!

File: src/blocks/gallery-masonry/controls.js [hunk 3]
```diff
@@ -8,11 +9,10 @@ import MediaFilterControl from '../../components/media-filter-control';
  */
 import { BlockControls } from '@wordpress/block-editor';
 
-const Controls = ( props ) => {
-	const { attributes } = props;
+const Controls = ( props, ) => {
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/controls.js:12]: rm trailing comma

File: babel.config.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "presets": [
+        "@wordpress/babel-preset-default"
+    ],
+    "plugins": [
+        "@babel/plugin-proposal-class-properties"
+    ]
+}
```
Comment (jlemay-godaddy) [inline babel.config.json:8]: New Line at EOF

File: src/blocks/gallery-masonry/edit.js [hunk 5]
```diff
@@ -2,268 +2,559 @@
...
+			ALLOWED_MEDIA_TYPES.some(
+				( mediaType ) => file.type?.indexOf( mediaType ) === 0
+			) || file.url?.indexOf( 'blob:' ) === 0
+		);
+	}
+
+	function updateImages( selectedImages ) {
+		const newFileUploads =
+			Object.prototype.toString.call( selectedImages ) ===
+			'[object FileList]';
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/edit.js:234]: First time I see something like that. Doesn't mean it's wrong

File: src/blocks/gallery-masonry/styles/editor.scss [hunk 6]
```diff
@@ -18,118 +17,119 @@
...
+				padding: 0;
+				width: 100%;
 			}
 
-			&:not(.mce-content-body) {
-				opacity: 0.6;
+			figcaption {
+				background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.7) 0, rgba($color: $black, $alpha: 0.3) 50%, transparent);
+				bottom: 0;
+				font-size: 13px;
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/styles/editor.scss:64]: We should try to use "em" or "rem" for font-sizes. It will scale better when zooming

File: src/blocks/gallery-masonry/styles/editor.scss [hunk 7]
```diff
@@ -18,118 +17,119 @@
...
-		// IE calculates this incorrectly, so leave it to modern browsers.
-		@supports (position: sticky) {
-			bottom: 4px;
-			left: 4px;
-			margin-top: -4px;
-			right: 4px;
-			width: calc(100% - 8px);
+		/* Masonry on tablets */
+		@media only screen and (min-width: 768px) and (max-width: 1023px) {
+			& {
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/styles/editor.scss:91]: I think (I might be wrong) you can remove lines 91 and 93. Same with 98 and 100

File: src/blocks/gallery-masonry/styles/style.scss [hunk 8]
```diff
@@ -1,44 +1,105 @@
-.wp-block-coblocks-gallery-masonry {
-	position: relative;
+// Import styles for rendering the static content of deprecated gallery versions.
+@import "../v1/deprecated.scss";
 
-	ul {
-		list-style: none !important;
-		padding: 0 !important;
+figure.wp-block-coblocks-gallery-masonry.masonry-grid {
+	column-fill: initial;
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/styles/style.scss:5]: I see a lot of code duplication in the editor.scss and style.scss files. Is the backend only loading the editor.scss and the frontend only loading the style.scss? I'm afraid we do a fix on the editor.scss and forget style.scss or something like that. How do we handle that usually?

File: src/blocks/gallery-masonry/styles/style.scss [hunk 9]
```diff
@@ -1,44 +1,105 @@
...
+			margin: 0;
+			max-width: 100%;
+			padding: 0;
+			width: 100%;
+		}
+
+		figcaption {
+			background: linear-gradient(0deg, rgba($color: $black, $alpha: 0.7) 0, rgba($color: $black, $alpha: 0.3) 50%, transparent);
+			bottom: 0;
+			font-size: 13px;
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/styles/style.scss:36]: see editor.scss

File: src/blocks/gallery-masonry/v1/block.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+	"name": "coblocks/gallery-masonry",
+	"category": "media",
+	"attributes": {
+		"gridSize": {
+			"type": "string",
+			"default": "xlrg"
+		}
+	}
+}
```
Comment (jlemay-godaddy) [inline src/blocks/gallery-masonry/v1/block.json:10]: NL at EOF

**Review by jlemay-godaddy — COMMENTED** [review]
Good job! Code review done. I'll test it IRL now :)

File: src/blocks/gallery-masonry/v1/edit.js [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+			[ `coblocks-animate ${ animation }` ]: animation,
+		}
+	);
+
+	return (
+		<>
+			{ isSelected &&
+			<Controls
+				{ ...props }
+			/>
```
Comment (jonathanbardo) [inline src/blocks/gallery-masonry/v1/edit.js:205]:
```suggestion
			<Controls { ...props } />
```

File: src/blocks/gallery-masonry/v1/edit.js [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+			}
+			{ isSelected &&
+			<Inspector
+				{ ...props }
+			/>
```
Comment (jonathanbardo) [inline src/blocks/gallery-masonry/v1/edit.js:210]: I don't think we need the extra spacing here.
```suggestion
			<Inspector { ...props } />
```

File: src/blocks/gallery-masonry/styles/editor.scss (hunk 6 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/styles/editor.scss:64]: That sounds good to me. My only hesitation is that a quick scan of the code reveals that we set captions on most galleries to `font-size: 13px` including stacked and offset galleries. These are similar styles in that way. I wonder if we should change them all at once?

File: src/blocks/gallery-masonry/styles/style.scss (hunk 8 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/styles/style.scss:5]: This was just oversight on my part most likely. Thank you for finding this. The editor.scss should load only in the editor whereas the style.scss should load in both the editor as well as the front end.

File: src/blocks/gallery-masonry/styles/editor.scss (hunk 7 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/styles/editor.scss:91]: Good call. This logic should exist only in the style.scss file and not here. This is duplicated code.

Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/styles/editor.scss:91]: Oh weird. I just tried removing these and the styles in the editor broke. I wonder if I didn't install the ones under styles.scss correctly. 🤔

Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/styles/editor.scss:94]: Oh, I got it now. I ended up having some duplicated code because we are targeting differing hierarchies of DOM selectors in the edit and the save markup. Because of this variation, the solution I went for was to include the same styles in both the editor.scss and style.scss file but using the applicable selector. I've updated these files to include a comment about the code duplication and why it exists so that we might improve it in the future.

File: src/blocks/gallery-masonry/v1/edit.js (hunk 12 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/v1/edit.js:210]: Fixed the spacing 👍

**Comment by AnthonyLedesma** [comment]
Pushed up fixes to address the formatting feedback. Also found out that the stacked jest tests were formatting incorrectly causing the results not to be correct. I fixed that here too.

File: src/blocks/gallery-masonry/controls.js (hunk 1 above)
Comment (AnthonyLedesma) [reply src/blocks/gallery-masonry/controls.js:1]: For posterity, I ended up re-adding the media-filter controls the masonry block with some refactors to extension logic. We now have a hierarchy of controls.
